### PR TITLE
More fixes

### PR DESCRIPTION
**Pull request by MatthijsBurgh**

### CODE REVIEW

**Comment by sizmailov** [comment]
# [Codecov](https://codecov.io/gh/sizmailov/pybind11-stubgen/pull/16?src=pr&el=h1) Report
> Merging [#16](https://codecov.io/gh/sizmailov/pybind11-stubgen/pull/16?src=pr&el=desc) into [master](https://codecov.io/gh/sizmailov/pybind11-stubgen/commit/7e6e59e44e85d9ff8d842842559e84b1abfaee71&el=desc) will **increase** coverage by `0.22%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/sizmailov/pybind11-stubgen/pull/16?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master      #16      +/-   ##
==========================================
+ Coverage   83.03%   83.25%   +0.22%     
==========================================
  Files           2        2              
  Lines         448      448              
==========================================
+ Hits          372      373       +1     
+ Misses         76       75       -1     
```



------

[Continue to review full report at Codecov](https://codecov.io/gh/sizmailov/pybind11-stubgen/pull/16?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/sizmailov/pybind11-stubgen/pull/16?src=pr&el=footer). Last update [7e6e59e...3dc3b39](https://codecov.io/gh/sizmailov/pybind11-stubgen/pull/16?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by sizmailov** [comment]
This fragment is same to one in #15. I think it would be better to refactor both and avoid repetition.

Tests are not required but would be appreciated.

**Comment by MatthijsBurgh** [comment]
Shared code is indeed better. Might be good to move this to `remove_signature` (Rename of that method would also be good in that case)

Do you agree @sizmailov ?

**Comment by sizmailov** [comment]
Yes, I've been thinking about a new class methods in common base class. 
Currently `remove_signature` does precisely what name says. I would write another one and reuse it instead of pushing more functionality in single function. It's better to have isolated responsibilities. 

If you want to proceed, do the most reasonable change you think of, update the PR and then we will see how it goes. 

(I'm currently in a rush, I almost have no spare time, so won't be able to do much here in next two-three weeks)

**Comment by MatthijsBurgh** [comment]
@sizmailov what do you think of these changes?

File: pybind11_stubgen/__init__.py [hunk 1]
```diff
@@ -221,6 +231,17 @@ def remove_signatures(docstring):  # type: (str) ->str
...
         return "\n".join(filter(lambda line: not re.match(signature_regex, line), lines))
 
+    @staticmethod
+    def filter_docstring(docstring):  # type: (str) ->str
+        docstring_new = StubsGenerator.remove_signatures(docstring)
+
+        # Reduce the double empty line to a single empty line
+        docstring_new = re.sub(r"\n\s*\n", "\n\n", docstring_new)
+        if docstring_new and re.match(r"^\s*$", docstring_new):
+            docstring_new = ""
```
Comment (sizmailov) [inline pybind11_stubgen/__init__.py]: Those lines needs to be in another function, named something like `strip_repeated_empty_lines`.
Then implementation of `sanitize_docstring` will be imperative an concise:

```python
docstring = StubsGenerator.remove_signatures(docstring)
docstring = StubsGenerator.remove_strip_repeated_empty_lines(docstring)
return docstring
```

File: pybind11_stubgen/__init__.py [hunk 2]
```diff
@@ -470,20 +480,21 @@ def strip_current_module_name(obj, module_name):
...
                 base_classes_list=", ".join(base_classes_list),
-                doc_string='\n' + self.INDENT + '"""{}"""'.format(self.doc_string) if self.doc_string else "",
+                doc_string='\n' + self.indent_lines('"""\n{}\n"""'.format(self.doc_string.strip("\n")))
+                            if self.doc_string else "",
             ),
         ]
         for f in self.methods:
             if f.name not in self.methods_blacklist:
-                result.extend(map(self.indent, f.to_lines()))
+                result.extend(map(self.indent_lines, f.to_lines()))
```
Comment (sizmailov) [inline pybind11_stubgen/__init__.py]: `map` is redundant now

File: pybind11_stubgen/__init__.py [hunk 3]
```diff
@@ -379,20 +390,19 @@ def parse(self):
...
 
-        docstring = self.remove_signatures(self.prop.__doc__)
-        docstring += "\n:type: {rtype}".format(rtype=self.signature.rtype)
+        docstring = self.filter_docstring(self.prop.__doc__)
+        docstring_prop = "\n\n".join([docstring, ":type: {rtype}".format(rtype=self.signature.rtype)])
 
         result = ["@property",
                   "def {field_name}(self) -> {rtype}:".format(field_name=self.name, rtype=self.signature.rtype),
-                  self.indent('"""{}"""'.format(docstring))]
+                  self.indent_lines('"""\n{}\n"""'.format(docstring_prop.strip("\n")))]
```
Comment (sizmailov) [inline pybind11_stubgen/__init__.py]: This operations on `\n` seems to be repeated as well and needs to be extracted to a new function too.

File: pybind11_stubgen/__init__.py [hunk 4]
```diff
@@ -117,6 +117,16 @@ def to_lines(self):  # type: () -> List[str]
     def indent(line):  # type: (str) -> str
         return StubsGenerator.INDENT + line
 
+    @staticmethod
+    def indent_lines(lines):
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply pybind11_stubgen/__init__.py]: `_indent` is now the actual implementation
`indent` does now handle also multiline strings
`indent_lines` converted also lists to strings, which was incorrect, so removed.

File: pybind11_stubgen/__init__.py (hunk 2 above)
Comment (MatthijsBurgh) [reply pybind11_stubgen/__init__.py]: `indent_lines` was wrong. So `self.indent` is back and `map` should stay.

File: pybind11_stubgen/__init__.py [hunk 5]
```diff
@@ -221,6 +227,26 @@ def remove_signatures(docstring):  # type: (str) ->str
 
         return "\n".join(filter(lambda line: not re.match(signature_regex, line), lines))
 
+    @staticmethod
+    def remove_repeated_empty_lines(docstring):
+        # Reduce the double empty line to a single empty line
+        return re.sub(r"\n\s*\n", "\n\n", docstring)
```
Comment (sizmailov) [inline pybind11_stubgen/__init__.py]: Turn comment to docstring

File: pybind11_stubgen/__init__.py [hunk 6]
```diff
@@ -221,6 +227,26 @@ def remove_signatures(docstring):  # type: (str) ->str
...
+
+    @staticmethod
+    def sanitize_docstring(docstring):  # type: (str) ->str
+        docstring = StubsGenerator.remove_signatures(docstring)
+        docstring = StubsGenerator.remove_repeated_empty_lines(docstring)
```
Comment (sizmailov) [inline pybind11_stubgen/__init__.py]: I doubt about removing repeated empty lines. The format of docstring is up to library author, probably we should limit our-self to trimming outer empty lines only, see https://www.python.org/dev/peps/pep-0257/#id18 

The docstring indentation can be managed in another PR

Comment (MatthijsBurgh) [reply pybind11_stubgen/__init__.py]: True, we shouldn't force our rules. The reason why I implemented it is by removing the signatures, we create these double empty lines.

We should actually fix that problem. So that we remove the signature line, including the first or previous empty line.

PS, I experienced two problems with the signature removing.
1. It can't handle any brackets in the signature, this could possibly happen in the default value representation.
2. It can't handle new lines in the signature. This is caused by splitting on "\n". Signatures are always separated by an empty line, the one we need to remove, so splitting on "\n\n" would maybe fix both problems in once.

Comment (MatthijsBurgh) [reply pybind11_stubgen/__init__.py]: Splitting on "\n\n" (and again joining on "\n\n"), does fix the problem of creating double empty lines, when remove signatures. Therefore there is no need to remove the double lines anymore, which would also change the original spacing in the remainder of the docstring.

File: pybind11_stubgen/__init__.py (hunk 5 above)
Comment (MatthijsBurgh) [reply pybind11_stubgen/__init__.py]: Function is removed.

**Comment by MatthijsBurgh** [comment]
The remaining issues:
1. Fix the regex to handle brackets in the signatures
2. Extend testing

If you agree with the current state, could you review it.

When  you have time in the a few weeks you can add the test. I could help with reviewing these.
I am not really experienced with regex, so could address that issue as well?

**Comment by sizmailov** [comment]
Testing and regex update should go to a separate PR. I'm inclined to merge this after function rename.
